keps/sig-node/20191118-cgroups-v2.md

```diff
@@ -0,0 +1,241 @@
+prevents Kubernetes from working as it is required to run with cgroups
+v1.
+
+## Goals
+
+This proposal aims to:
+
+*   Add support for cgroups v2 to the Kubelet
+
+## Non-Goals
+
+*	Expose new cgroup2-only features
+*	Dockershim
+*	Plugins support
+
+## User Stories
+
+*   The Kubelet can run on a host using either cgroups v1 or v2.
+*   Have features parity between cgroup v2 and v1.
+
+## Implementation Details
+
+## Proposal
+
+The proposal is to implement cgroups v2 in two different phases.
+
+The first phase ensures that any configuration file designed for
+cgroups v1 will continue to work on cgroups v2.
+
+The second phase requires changes through the entire stack, including
+the OCI runtime specifications.
+
+At startup the Kubelet detects what hierarchy the system is using.  It
+checks the file system type for `/sys/fs/cgroup` (the equivalent of
+`stat -f --format '%T' /sys/fs/cgroup`).  If the type is `cgroup2fs`
+then the Kubelet will use only cgroups v2 during all its execution.
+
+The current proposal doesn't aim at deprecating cgroup v1, that must
+still be supported through the stack.
+
+Device plugins that require v2 enablement are out of the scope for
+this proposal.
+
+### Dependencies on OCI and container runtimes
+
+In order to support features only available in cgroups v2 but not in
+cgroups v1, the OCI runtime specs must be changed.
+
+New features that are not present in cgroup v1 are out of the scope
+for this proposal.
+
+The dockershim implementation embedded in the Kubelet won't be
+supported on cgroup v2.
+
+### Current status of dependencies
+
+- CRI-O+crun: support cgroups v2
+
+- runc support for cgroups v2 is work in progress [current status](#current-cgroups-usage-and-the-equivalent-in-cgroups-v2)
+
+- containerd: [https://github.com/containerd/containerd/issues/3726](https://github.com/containerd/containerd/issues/3726)
+
+- Moby: [https://github.com/moby/moby/pull/40174](https://github.com/moby/moby/pull/40174)
+
+- OCI runtime spec: TODO
+
+- cAdvisor already supports cgroups v2 ([https://github.com/google/cadvisor/pull/2309](https://github.com/google/cadvisor/pull/2309))
+
+## Current cgroups usage and the equivalent in cgroups v2
+
+|Kubernetes cgroups v1|Kubernetes cgroups v2 behavior|
+|---|---|
+|CPU stats for Horizontal Pod Autoscaling|No .percpu cpuacct stats.|
+|CPU pinning based on integral cores|Cpuset controller available|
+|Memory limits|Not changed, different naming|
+|PIDs limits|Not changed, same naming|
+|hugetlb|Added to linux-next, targeting Linux 5.6|
+
+### cgroup namespace
+
+A cgroup namespace restricts the view on the cgroups.  When
+unshare(CLONE_NEWCGROUP) is done, the current cgroup the process
+resides in becomes the root.  Other cgroups won't be visible from the
+new namespace.  It was not enabled by default on a cgroup v1 system as
+older kernel lacked support for it.
+
+Privileged pods will still use the host cgroup namespace so to have
```

Comment:
Let's make this customizable, especially for nested containers:
- https://github.com/kubernetes/enhancements/pull/5715